plugins/shrs_completion/src/setup.rs

```diff
@@ -0,0 +1,118 @@
+use rhai::{Array, CustomType, Dynamic, Engine, ImmutableString, Scope, Shared, TypeBuilder};
+use shrs::prelude::*;
+fn default_completion(completion: String) -> Completion {
+    Completion {
+        add_space: true,
+        display: None,
+        completion,
+        replace_method: ReplaceMethod::Replace,
+        comment: None,
+    }
+}
+fn default_completion_with_comment(completion: String, comment: String) -> Completion {
+    Completion {
+        add_space: true,
+        display: None,
+        completion,
+        replace_method: ReplaceMethod::Replace,
+        comment: Some(comment),
+    }
+}
+
+fn new_completion(
+    add_space: bool,
+    display: String,
+    completion: String,
+    replace_method: ReplaceMethod,
+    comment: String,
+) -> Completion {
+    Completion {
+        add_space,
+        display: Some(display),
+        completion,
+        replace_method,
+        comment: Some(comment),
+    }
+}
+fn append_method() -> ReplaceMethod {
+    ReplaceMethod::Append
+}
+fn replace_method() -> ReplaceMethod {
+    ReplaceMethod::Replace
+}
+fn get_line(ctx: &mut CompletionCtx) -> Vec<String> {
+    ctx.line.clone()
+}
+fn cmd_name(ctx: &mut CompletionCtx) -> Dynamic {
+    if let Some(n) = ctx.cmd_name() {
+        return n.into();
+    }
+    Dynamic::UNIT
+}
+fn cur_word(ctx: &mut CompletionCtx) -> Dynamic {
+    if let Some(w) = ctx.cur_word() {
+        return w.into();
+    }
+    Dynamic::UNIT
+}
+fn arg_num(ctx: &mut CompletionCtx) -> i64 {
+    ctx.arg_num() as i64
+}
+fn df(arr: Array) -> Array {
+    let c: Vec<String> = arr
+        .iter()
+        .map(|s| s.clone().into_string().unwrap())
+        .collect();
+    let g = default_format(c);
+    g.iter().map(|f| Dynamic::from(f.clone())).collect()
+}
+fn dfc(arr: Array) -> Array {
+    let c: Vec<(String, String)> = arr
+        .iter()
+        .map(|s| s.clone().into_typed_array::<String>().unwrap())
+        .map(|a| (a.first().unwrap().clone(), a.last().unwrap().clone()))
+        .collect();
+    let g = default_format_with_comment(c);
+    g.iter().map(|f| Dynamic::from(f.clone())).collect()
+}
+fn filename_completions(ctx: CompletionCtx) -> Array {
```

Comment:
don't need to be super descriptive but some docstrings for these would be nice, especially since `df` and `dfc` are a bit more cryptic for function names